### PR TITLE
Issue #740 Fix incorrect arguments provided to the clip_box method of the well package

### PR DESCRIPTION
**Pull request by Manangka**
In this commit the wrong arguments provided to the well package is addressed. 

The Well package is a grid agnostic package. That means that the layer_min and layer_max arguments from the clip_box method has no meaning to it. 

To still conform to this method signature the top and bottom of a grid is provided. Those are then used to retrieve a z_min and z_max

### CODE REVIEW

File: imod/tests/test_mf6/test_mf6_wel.py [hunk 1]
```diff
@@ -137,35 +138,51 @@ def test_to_mf6_pkg__high_lvl_transient(basic_dis, well_high_lvl_test_data_trans
     np.testing.assert_equal(mf6_ds["rate"].values, rate_expected)
 
 
-def test_clip_box__high_lvl_stationary(well_high_lvl_test_data_stationary):
+@pytest.mark.parametrize(
```
Comment (luitjansl) [inline imod/tests/test_mf6/test_mf6_wel.py:141]: it would be nice to add test cases where for example layer_min is not None, but layer_max is. And vice versa. Or is that combination not allowed?
Also the parameters time_min and time_max could be included in the testing

File: imod/mf6/package.py [hunk 2]
```diff
@@ -401,16 +403,18 @@ def __to_datetime(self, time, use_cftime):
...
+        x_min: Optional[float] = None,
+        x_max: Optional[float] = None,
+        y_min: Optional[float] = None,
+        y_max: Optional[float] = None,
+        top: Optional[GridDataArray] = None,
+        bottom: Optional[GridDataArray] = None,
+        state_for_boundary: Optional[GridDataArray] = None,
+    ) -> Package:
         """
         Clip a package by a bounding box (time, layer, y, x).
```
Comment (luitjansl) [inline imod/mf6/package.py:419]: Maybe the docstring should be expanded a little to explain that if layer_min is present, bottom should also be present, and if layer_max is present, top should be present.

File: imod/mf6/package.py [hunk 3]
```diff
@@ -401,16 +403,18 @@ def __to_datetime(self, time, use_cftime):
...
-        time_max=None,
-        layer_min=None,
-        layer_max=None,
-        x_min=None,
-        x_max=None,
-        y_min=None,
-        y_max=None,
-        state_for_boundary=None,
-    ) -> "Package":
+        time_min: Optional[str] = None,
```
Comment (luitjansl) [inline imod/mf6/package.py:406]: is str the type of t_min? Isn't there a more precise type that could be used? Maybe here some input checking can also be added to make sure t_min has a string value that makes sense.

**Review by luitjansl — COMMENTED** [review]
looks good, some comments

File: imod/mf6/wel.py [hunk 4]
```diff
@@ -237,6 +243,9 @@ def clip_box(
 
         ds = new.dataset
 
+        z_max = None if layer_max is None else top.isel(layer=layer_max)
```
[the remark this answers is not in the document]
Comment (JoerivanEngelen) [reply imod/mf6/wel.py:246]: Before this changeset, we expected ``z_max`` and ``z_min`` as a scalar. 
This makes it easy to make simple selections in the table of wells.
This was not feature complete, more a pragmatic workaround so that I not had to deal with locating wells in the grid.

However, top and bottom most of the time are grids. ``z_min``/``z_max`` therefore cannot be expected to be scalar anymore, which means well points have to be located in the grid to do the selecting (at least if ``top`` and ``bottom`` are provided as grids). 
I think we can use ``imod.prepare.wells.locate_wells`` for this. With this we can discard wells easily which are outside top and bottom.

Furthermore: ``top`` can also be provided as scalar by the user, I'm not sure if that perpetuates to the ``clip_box`` method here, or that ``top`` has a got a layer dim assigned to it in the mean time.

File: imod/tests/test_mf6/test_mf6_wel.py (hunk 1 above)
Comment (JoerivanEngelen) [reply imod/tests/test_mf6/test_mf6_wel.py:141]: I see we are now also using testdata where bottom, top are not grids. In the fixtures: 

```python
@pytest.fixture
def parameterizable_basic_dis(request):
...
    top = xr.DataArray(z[:-1], coords={"layer": layers})
    bottom = xr.DataArray(z[1:], coords={"layer": layers})
...
```

This is possible, but doesn't cover most of the use cases. Most of the applications have ``top`` and ``bottom`` specified as grids. We should cover that in our tests as well.

**Review by JoerivanEngelen — CHANGES_REQUESTED** [review]
2 comments, I think we have to carefully look what dimensions ``top`` and ``bottom`` are/should be provided with here. As layers or 3Dgrid ? And in case of ``top`` as scalar or as 2D grid?

File: imod/mf6/wel.py [hunk 5]
```diff
@@ -220,31 +231,50 @@ def clip_box(
...
+        z_min = None if layer_min is None else bottom.isel(layer=layer_min)
+
+        if z_max:
+            ds["screen_top"] = ds["screen_top"].clip(None, z_max)
+        if z_min:
+            ds["screen_bottom"] = ds["screen_bottom"].clip(z_min, None)
```
Comment (JoerivanEngelen) [inline imod/mf6/wel.py:266]: I don't think this call to ``clip`` works in the case that ``top`` and ``bottom`` is a grid, since wells are not defined as a grid.

File: imod/tests/test_mf6/test_mf6_wel.py [hunk 6]
```diff
@@ -137,35 +140,120 @@ def test_to_mf6_pkg__high_lvl_transient(basic_dis, well_high_lvl_test_data_trans
     np.testing.assert_equal(mf6_ds["rate"].values, rate_expected)
 
 
-def test_clip_box__high_lvl_stationary(well_high_lvl_test_data_stationary):
+class ClipBoxCases:
```
Comment (JoerivanEngelen) [inline imod/tests/test_mf6/test_mf6_wel.py:143]: I like these cases! I think you should add some cases where ``top`` and ``bottom`` are provided as grids to see what I'm talking about in my other comment ;-)

FYI: In the ``parameterizable_basic_dis`` fixture, ``top`` and ``bottom`` are not created as grids:

```python
    top = xr.DataArray(z[:-1], coords={"layer": layers})
    bottom = xr.DataArray(z[1:], coords={"layer": layers})
```

**Review by JoerivanEngelen — CHANGES_REQUESTED** [review]
The changes made are an improvement, but I think you should add some test cases with ``top`` and ``bottom`` as grids to see what I'm talking about ;-)

File: imod/mf6/wel.py [hunk 7]
```diff
@@ -220,31 +231,63 @@ def clip_box(
...
+        if (z_min is not None) and ("x" in z_min.coords or "y" in z_min.coords):
+            z_min = z_min.sel(x=ds["x"], y=ds["y"], method="nearest").drop_vars(
+                lambda x: x.coords
+            )
+
+        if z_max is not None:
+            ds["screen_top"] = ds["screen_top"].clip(None, z_max)
+        if z_min is not None:
+            ds["screen_bottom"] = ds["screen_bottom"].clip(z_min, None)
+
```
Comment (JoerivanEngelen) [inline imod/mf6/wel.py:280]: Already better, but there are still some issues: 

1. This doesn't work for unstructured grids, these do not have "x" and "y" in coords, so then the if statements fails, as well as the call to ``.sel``, this should be ``.ugrid.sel``. Luckily we don't have to reinvent the wheel. We already wrote ``imod.typing.grid.is_spatial_2D`` for this and ``imod.select.points_values``.
2.  ``z_min`` and ``z_max`` can be None, but can also be first a grid and then be converted to a 1d array. I think it is best to keep ``z_min``, ``z_max`` to something that is accepted by ``values_within_range``, so not as grid.

In conclusion: 

```suggestion
        z_max_grid = None if layer_max is None else top.isel(layer=layer_max)
        z_min_grid = None if layer_min is None else bottom.isel(layer=layer_min)

        # if z is a grid select the the values at the well locations and drop the dimensions
        if (z_max_grid is not None) and is_spatial_2D(z_max_grid):
            z_max =  imod.select.points_values(z_max_grid, x=ds["x"].values, y=ds["y"].values, out_of_bounds="ignore")
            )
        else: 
            z_max = None

        if ( z_min_grid is not None) and is_spatial_2D(z_min_grid):
            z_min =  imod.select.points_values(z_min_grid, x=ds["x"].values, y=ds["y"].values, out_of_bounds="ignore")
            )
        else: 
            z_min = None
```

The call to ``is_spatial_2D(z_max_grid)`` is probably overdoing things, so you can probably remove that one.

File: imod/tests/test_mf6/test_mf6_wel.py [hunk 8]
```diff
@@ -137,35 +141,141 @@ def test_to_mf6_pkg__high_lvl_transient(basic_dis, well_high_lvl_test_data_trans
...
+        top = 0.0
+        clip_arguments = {"layer_max": 2, "bottom": bottom, "top": top}
+
+        expected_dims = {"index": 4, "species": 2}
+        return clip_arguments, expected_dims, does_not_raise()
+
+    @staticmethod
+    def case_clip_top_is_non_layered_griddataarray(parameterizable_basic_dis):
+        _, top, bottom = parameterizable_basic_dis
+        top = top.isel(layer=0).drop_vars("layer")
```
Comment (JoerivanEngelen) [inline imod/tests/test_mf6/test_mf6_wel.py:192]: I don't see top here defined as griddataarray, missing the call to ``broadcast_to_full_domain here``.

**Review by JoerivanEngelen — CHANGES_REQUESTED** [review]
Looks better, I have some remarks/suggestions though!

File: imod/tests/test_mf6/test_mf6_wel.py [hunk 9]
```diff
@@ -137,35 +141,141 @@ def test_to_mf6_pkg__high_lvl_transient(basic_dis, well_high_lvl_test_data_trans
     np.testing.assert_equal(mf6_ds["rate"].values, rate_expected)
 
 
-def test_clip_box__high_lvl_stationary(well_high_lvl_test_data_stationary):
+class ClipBoxCases:
```
Comment (JoerivanEngelen) [inline imod/tests/test_mf6/test_mf6_wel.py:144]: Looks better! I'm still missing a test with an unstructured grid, you can easily convert a structured grid to an unstructured grid with: 
``idomain_ugrid = xu.UgridDataArray.from_structured(idomain)``

File: imod/mf6/wel.py (hunk 7 above)
Comment (Manangka) [reply imod/mf6/wel.py]: This worked like a charm. I added 2 unit tests for unstructured grids and with your code they both pass
